PICARD-1716: Add script functions for strings, multi-value variables, and loops

rdswift opened the pull request:
<!--
    Hello! Thanks for submitting a pull request to MusicBrainz Picard. We
    appreciate your time and interest in helping our project!

    Use this template to help us review your change. Not everything is required,
    depending on your change. Keep or delete what is relevant for your change.
    Remember that it helps us review if you give more helpful info for us to
    understand your change.

    Ensure that you've read through and followed the Contributing Guidelines, in
    [CONTRIBUTING.md](https://github.com/metabrainz/picard/blob/master/CONTRIBUTING.md).
-->

# Summary
Add new functions to the scripting library.
<!--
    Update the checkbox with an [x] for the type of contribution you are making.
-->

* This is a…
    * [ ] Bug fix
    * [x] Feature addition
    * [ ] Refactoring
    * [ ] Minor / simple change (like a typo)
    * [ ] Other
* **Add new functions as:**
    * 'find' substring within string
    * string 'reverse'
    * 'substr' to extract a substring from a string
    * 'get_multi_item' to get an element from a multi-value variable
    * 'foreach' to loop through each item in a multi-value variable
    * 'for' loop
    * 'while' loop

# Problem

This additional functionality will allow users to easily access individual elements of multi-value variables, and perform repetitive code within loops that can be dynamically constructed.  Additional string functions allow searching for and extracting substrings.  Reverse can be used to find the last occurrence of a substring rather than just the first.
<!--
    Anything that helps us understand why you are making this change goes here.
    What problem are you trying to fix? What does this change address?
-->

* JIRA ticket (_optional_): PICARD-1716
<!--
    Please make sure you prefix your pull request title with 'PICARD-XXX' in order
    for our ticket tracker to link your pull request to the relevant ticket.
-->



# Solution

<!--
    The details of your change. Talk about technical details, considerations, or
    other interesting points. If you have a lot to say, be more detailed in this
    section.
-->

Suggested changes to the Scripting page of the Picard Website include:



**$find(haystack,needle)**

Find the location of the first occurrence of one string within another. Returns the zero-based index of the first occurrance of needle in haystack, or -1 if needle was not found.  Examples:

```
$findstr(abcdef,cde) returns 2
$findstr(abcdef,) returns 0
$findstr(abcdef,cdf) returns -1
```

**$reverse(text)**

Returns text in reverse order.  Example:

```
$reverse(abcde) returns 'edcba'
```

**$substr(text,start_index,end_index)**

Extract a specified portion of a string.  Returns the substring beginning with the character at the start index, up to (but not including) the character at the end index.  The first character is at index number 0.  If the start index is left blank, it defaults to the first character in the string.  If the end index is left blank, it defaults to the number of characters in the string.  If either index is negative, it is subtracted from the total number of characters in the string to provide the index used.  Examples:

```
$substr(abcde,1,2) returns 'b'
$substr(abcde,,2) returns 'ab'
$substr(abcde,1,) returns 'bcde'
$substr(abcde,,) returns 'abcde'
$substr(abcde,-2,) returns 'de'
$substr(abcde,-4,-2) returns 'bc'
```

**$get_multi_item(multi,item_index)**

Returns the value of the item at the specified index in the multi-value variable.  Index values are zero-based.  Example:

```
$get_multi_item(a; b; c; d; e,2) returns 'c'
```

**$foreach(multi,loop_code,_loop_value="_loop_value",_loop_count="_loop_count")**

Iterates over each element found in the specified multi-value variable, executing the specified code. For each loop, the element value is first stored in the tag specified by _loop_value and the count is stored in the tag specified by _loop_count.  This allows the element or count value to be accessed within the code script.  Example:

```
$foreach(isrc; year; comment; artists,$unset(%_loop_value%))
```

**$for(name,start,end,increment,loop_code)**

Standard 'for' loop.  Sets the value of name to the start value (integer) and loops through the specified code until the value of name is greater than or equal to the end value (integer).  At the end of each loop, the value of name is incremented by the specified increment value (integer).  Example:

```
$set(testing,Testing)
$for(_count,1,4,1,$set(testing,%testing% %_count%...))
```

**$while_loop(condition,loop_code)**

Standard 'while' loop.  Loops through the loop_code until the condition is no longer True.  Example:

```
$set(testing,Echo)
$while($lt($len(%testing%,20)),$set(testing,%testing% echo...))
```


# Action

<!--
    Other than merging your change, do you want / need us to do anything else
    with your change? This could include reviewing a specific part of your PR.
-->

phw commented on picard/script.py:
```diff
@@ -910,6 +910,143 @@ def func_is_video(parser):
+               characters in the string to provide the index used.
+    """
+    _start = int(start_index) if start_index else None
+    _end = int(end_index) if end_index else None
```
This must catch `ValueError` in case the int conversion fails. I'd say we treat the value throwing the exception should just be considered `None`.

rdswift replied:
Updated to include the ValueError check.

phw commented on picard/script.py:
```diff
@@ -967,3 +1104,10 @@ def func_is_video(parser):
+register_script_function(func_find_str, "find")
+register_script_function(func_reverse_str, "reverse")
+register_script_function(func_substr, "substr")
+register_script_function(func_get_multi_item, "get_multi_item", eval_args=False)
```
Could we just call it `get_multi` or maybe `get_item`?

zas replied:
`get_multi` has my preference, and we can rename `func_*_multi_item` to `func_*_multi` as well.

rdswift replied:
Changed to $get_multi and func_get_multi.

phw commented on picard/script.py:
```diff
@@ -910,6 +910,143 @@ def func_is_video(parser):
+        str -- Returns value of the item at the specified index in the multi-value variable.  Index values are zero-based.
+    """
+    if item_index:
+        _index = int(item_index.eval(parser))
```
Again needs to fetch `ValueError`

rdswift replied:
Check added.

phw commented on picard/script.py:
```diff
@@ -910,6 +910,143 @@ def func_is_video(parser):
+    if name and start and end and increment and loop_code:
+        _count = int(start.eval(parser))
+        _end = int(end.eval(parser))
+        _inc = int(increment.eval(parser))
```
`ValueError` ;)

rdswift replied:
Check added.

zas commented on picard/script.py:
```diff
@@ -910,6 +910,143 @@ def func_is_video(parser):
+        return ''
+
+
+def func_foreach_multi_item(parser, multi, loop_code, _loop_value="_loop_value", _loop_count="_loop_count", separator=MULTI_VALUED_JOINER):
```
Wrap arguments?

rdswift replied:
I removed the option for different loop tag names (because I couldn't think of a case where alternate loop tag names would be necessary), and wrapped the rest.

Sophist-UK commented:
@rdswift Bob - just a reminder to update the script function reference on the Picard Web Site repo once this has merged.

phw commented:
We should also have test cases for the new functions in `test_script.py`

samj1912 commented on picard/script.py:
```diff
@@ -910,6 +910,143 @@ def func_is_video(parser):
+
+    Arguments:
+        multi {multi-value} -- The multi-value from which the item is to be retrieved.
+        item_index {int} -- The zero-based index of the item to be retrieved.
```
How does an `int` type have an `eval` method?

samj1912 replied:
Let's change the type from `int` to the right Python type. (ScriptVariable/Function)

rdswift replied:
I actually removed the type designations altogether, because they didn't seem to be broken out in the example in the style guide you linked.

samj1912 commented on picard/script.py:
```diff
@@ -910,6 +910,143 @@ def func_is_video(parser):
+        loop_code {str} -- Script code to be processed on each iteration.
+    """
+    if name and start and end and increment and loop_code:
+        _count = int(start.eval(parser))
```
nitpick: Can we remove the `_` prepend from the variables?

rdswift replied:
Done.

zas commented on picard/script.py:
```diff
@@ -910,6 +910,143 @@ def func_is_video(parser):
+    _multi = _get_multi_values(parser, multi, separator)
+    for value in _multi:
+        loop_count += 1
+        func_set(parser, _loop_count, str(loop_count))
```
`loop_count` and `_loop_count` ?
Using '_' prefix is mainly for private functions, or unused variables (though that's just a convention).
Also you may want to use `enumerate()`.

rdswift replied:
Underscores removed.  I'll look at `enumerate()` (didn't notice that comment until after I submitted my update)..

rdswift replied:
Switched to use `enumerate()`.

zas commented on picard/script.py:
```diff
@@ -910,6 +910,143 @@ def func_is_video(parser):
+        end {int} -- The value at which the loop exits.
+        increment {int} -- The value by which the loop count variable is incremented for each loop.
+        loop_code {str} -- Script code to be processed on each iteration.
+    """
```
This function may not be that useful, it looks over complicated.

rdswift replied:
Yeah, I only included it because there is a `for()` loop in Python.  I agree that everything that the `for()` loop provides can be done using the `while()` loop so I've removed the `for()` loop from the code.

rdswift replied:
> Yeah, I only included it because there is a for() loop in Python.

Wrong.  I've been doing a lot of work in PHP lately, so that must have been what I was thinking of.  In any event, it's gone now.

phw replied:
I guess Python not having this is what made me frown on  the for loop :D

zas commented on picard/script.py:
```diff
@@ -910,6 +910,143 @@ def func_is_video(parser):
+        condition {str} -- Script code to check before each iteration through the loop.
+        loop_code {str} -- Script code to be processed on each iteration.
+    """
+    runaway_check = 1000
```
You should move this inside `if` block.

rdswift replied:
Right.  Done.

rdswift commented on picard/script.py:
```diff
@@ -910,6 +910,143 @@ def func_is_video(parser):
+    """Find the location of the first occurrence of one string within another.
+
+    Arguments:
+        haystack {str} -- The string to search
```
Docstrings updated.  Tried to follow the guide that you linked.

rdswift commented on picard/script.py:
```diff
@@ -910,6 +910,143 @@ def func_is_video(parser):
+               If either index is negative, it is subtracted from the total number of
+               characters in the string to provide the index used.
+    """
+    _start = int(start_index) if start_index else None
```
No reason at all.  They should all have the leading underscores removed now.

rdswift commented:
I've removed the redundant and complicated `for()` loop, added tests for all of the new functions, and added in a first draft of a `map()` function.  Thanks for all the helpful feedback.

zas commented on picard/script.py:
```diff
@@ -910,6 +910,159 @@ def func_is_video(parser):
+    """
+    loop_count = 0
+    multi_value = _get_multi_values(parser, multi, separator)
+    for index, value in enumerate(multi_value):
```
You can use `enumerate()` optional argument:

```python
 for loop_count, value in enumerate(multi_value, 1):
```

and get rid of `loop_count = 0` and `loop_count = index + 1`

rdswift replied:
That makes perfect sense.  Sorry I missed that.  I'll change it.

zas commented on picard/script.py:
```diff
@@ -910,6 +910,159 @@ def func_is_video(parser):
+
+    Arguments:
+        multi: The multi-value to be iterated.
+        loop_code: Script code to be processed on each iteration.
```
It has 4 arguments, but only 2 are documented here.

rdswift replied:
Should I remove all the arguments from the docstrings instead?  I see that none of the other functions include the arguments list.

zas replied:
Nope, better have more documentation

zas commented on picard/script.py:
```diff
@@ -910,6 +910,159 @@ def func_is_video(parser):
+    loop_count = 0
+    multi_value = _get_multi_values(parser, multi, separator)
+    for index, value in enumerate(multi_value):
+        loop_count = index + 1
```
Same remark: `enumerate(multi_value, 1)` and `multi_value[loop_count - 1] = ...` will simplify code a bit.

rdswift replied:
Will change.

zas commented on picard/script.py:
```diff
@@ -910,6 +910,159 @@ def func_is_video(parser):
+    """Standard 'while' loop.  Also includes a runaway check to limit the maximum number of iterations.
+
+    Arguments:
+        condition: Script code to check before each iteration through the loop.
```
First argument isn't documented

zas commented on picard/script.py:
```diff
@@ -910,6 +910,159 @@ def func_is_video(parser):
+
+    Arguments:
+        multi: The multi-value to be iterated.
+        loop_code: Script code to be processed on each iteration.
```
Missing documentation for 2 arguments: `parser` and `separator`

zas commented on picard/script.py:
```diff
@@ -910,6 +910,159 @@ def func_is_video(parser):
+    try:
+        return str(multi_var[index])
+    except IndexError:
+        return ''
```
I think it could be rewritten to a simpler form:

```python
    if not item_index:
        return ''
    try:
        index = int(item_index.eval(parser))
        multi_var = _get_multi_values(parser, multi, separator)
        return str(multi_var[index])
    except (ValueError, IndexError):
        return ''
```

rdswift replied:
Much cleaner!  I'll make the change.

rdswift commented:
A few more questions...

Now that we've removed the `for()` function, should we rename the new `foreach()` to just `for()` to keep it in line with `for()` in Python?

When returning the new multi-value in the `map()` function, it it is called with a separator override should it use the standard separator in the return string or should it use the override separator?

Is there any value in adding two more functions:  `$join(multi,join_phrase,separator='; ')` to return a string of the multi elements joined by the join phrase; and `$sub_multi(multi,start,end,separator='; ')` to return a slice of the original multi?  This would pretty much round out the functionality set.

Thanks.

zas commented on picard/script.py:
```diff
@@ -910,6 +910,163 @@ def func_is_video(parser):
+
+    Returns the updated multi-value variable.
+    """
+    loop_count = 0
```
You don't need to initialize `loop_count` here since it is done in the for loop.

rdswift replied:
Crap!  I missed that.  Thanks.

zas commented:
> Now that we've removed the `for()` function, should we rename the new `foreach()` to just `for()` to keep it in line with `for()` in Python?

You mean `foreach_multi` vs `for_multi` ?
I'd say keep it as is (`foreach_multi`).

> 
> When returning the new multi-value in the `map()` function, if it is called with a separator override should it use the standard separator in the return string or should it use the override separator?
> 

Code currently returns string using  override separator, AFAIK, it seems correct to me.


> Is there any value in adding two more functions: `$join(multi,join_phrase,separator='; ')` to return a string of the multi elements joined by the join phrase; and `$sub_multi(multi,start,end,separator='; ')` to return a slice of the original multi? This would pretty much round out the functionality set.

If you think they can be useful, why not. Though I'd do it in another PR, and perhaps name them using suffix `_multi`: `join_multi()` and `slice_multi()` (`sub` is misleading imho).
